node-wasm/src/ports.rs

```diff
@@ -0,0 +1,314 @@
+use js_sys::{Array, Function, Reflect};
+use serde::Serialize;
+use serde_wasm_bindgen::{from_value, to_value, Serializer};
+use tokio::select;
+use tokio::sync::{mpsc, Mutex};
+use tracing::{error, info, trace};
+use wasm_bindgen::closure::Closure;
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::{JsCast, JsValue};
+use wasm_bindgen_futures::spawn_local;
+use web_sys::{MessageEvent, MessagePort};
+
+use crate::commands::{NodeCommand, WorkerResponse};
+use crate::error::{Context, Error, Result};
+
+// Instead of just supporting communicaton with just `MessagePort`, allow using any object which
+// provides compatible interface
+#[wasm_bindgen]
+extern "C" {
+    pub type MessagePortLike;
+
+    #[wasm_bindgen (catch , method , structural , js_name = postMessage)]
+    pub fn post_message(this: &MessagePortLike, message: &JsValue) -> Result<(), JsValue>;
+
+    #[wasm_bindgen (catch , method , structural , js_name = postMessage)]
+    pub fn post_message_with_transferable(
+        this: &MessagePortLike,
+        message: &JsValue,
+        transferable: &JsValue,
+    ) -> Result<(), JsValue>;
+}
+
+impl From<MessagePort> for MessagePortLike {
+    fn from(port: MessagePort) -> Self {
+        JsValue::from(port).into()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct ClientId(usize);
+
+struct ClientConnection {
+    port: MessagePortLike,
+    _onmessage: Closure<dyn Fn(MessageEvent)>,
+}
+
+impl ClientConnection {
+    fn new(
+        id: ClientId,
+        object: JsValue,
```

Comment:
is object a port? can you add some meaningful name here?

Reply:
My idea was to accept just an object and figure out whether the interface is ok with `Reflect`, to be most compatible. `prepare_message_port` later sanitizes the input and gives you `port` back.

Reply:
so maybe `maybe_port`? just to help me understand :p

Reply:
`maybe_port` is confusing. what about `raw_port`?

Reply:
I'm trying to capture the spirit of "something that is a port or looks enough like a port for us to use it". I'm not sure about `maybe_port` or `raw_port` as they miss the ducktyping aspect of the situation. 
How about `maybe_port` (which coincidentally matches the `MessagePortLike` type).